stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeRepl.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2014-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+package com.facebook.stetho.inspector.console.js;
+
+import com.facebook.stetho.common.LogUtil;
+import com.facebook.stetho.inspector.console.RuntimeRepl;
+import org.mozilla.javascript.Context;
+import org.mozilla.javascript.ScriptableObject;
+
+class JavaScriptRuntimeRepl implements RuntimeRepl {
+
+  private final ScriptableObject mJsScope;
+
+  JavaScriptRuntimeRepl(ScriptableObject scope) {
+    mJsScope = scope;
+  }
+
+  @Override
+  public Object evaluate(String expression) throws Throwable {
+      LogUtil.i("js) eval expression: %s", expression);
+
+      Object result;
+      final Context jsContext = enterJsContext();
+      try {
+        result = jsContext.evaluateString(mJsScope, expression, "chrome", 1, null);
+      } finally {
+        Context.exit();
+      }
+
+      LogUtil.i("js) Result: %s; type: %s", result, result == null ? null : result.getClass().getName());
+      return Context.jsToJava(result, Object.class);
+  }
+
+  /**
+   * Setups a proper javascript context so that it can run javascript code properly under android.
+   * For android we need to disable bytecode generation since the android vms don't understand JVM bytecode.
+   * @return a proper javascript context
+   */
+  static Context enterJsContext() {
+    LogUtil.i("js) init JS context");
+    final Context jsContext = Context.enter();
+
+    // If we cause the context to throw a runtime exception from this point
+    // we need to make sure that exit the context.
+    try {
+      jsContext.setLanguageVersion(Context.VERSION_1_8);
+
+      // We can't let Rhino to optimize the JS and to use a JIT because it would generate JVM bytecode
+      // and android runs on DEX bytecode. Instead we need to go in interpreted mode.
+      jsContext.setOptimizationLevel(-1);
+    }
+    catch (RuntimeException e) {
+      Context.exit();
```

Comment:
We definitely need to enter the context each time since this is running on a different thread. The context needs to be enter each time that we run the JS code in a new thread. Without this each js eval call will throw an exception.

Think of it as in android when we operate on a UI thread object (ex: label.setText("foo")). When we do that android checks that the running thread is the main UI thread if not it throws a runtime exception.

Rhino does the same, it wants operations done through the interpreter form a different thread to be done by grabbing a context.

Github doesn't give me more context in the snippet but from what i remember this is the global js call that i have where I tell that i want to bytecode optimizations and to set the language to JS 1.8, the thing is that if one of these methods fails I need to ensure that I do exit the context.

The caller's pattern is:

```java
Context = enterJsContext(); // if this fails with a runtime error I need someone to close the context, but we're not in a try / finally yet
try {
  // use the context
}
finally {
  context.exit()
}
```

If I fail to build the context in enterJsContext() i still need to exit it as long as it is created. This is why inside enterJsContext() there's a catch block that exits the context.

I'll add more documentation for this.